### DISCUSSION TITLE
Implementing a Safety Score-Based Routing Strategy in Valhalla

### DISCUSSION THREAD

**Discussion by alz86** [valhalla/valhalla#4240] [Q&A] [answered]
Hello everybody,

I am currently involved in a project aimed at suggesting routes considering a given route's perceived safety. The primary audience is women looking to navigate their surroundings with an additional layer of safety in mind.

Our approach is to utilize OpenStreetMap (OSM) data, supplemented with additional data sources, to assign a "safety score" to each street (reported crime rates, presence of street lighting, etc.). 

The goal is to integrate this data into OSM and subsequently alter Valhalla's routing engine to consider this safety score as a penalty when determining optimal routes.

Before delving deep into implementation, I'd like to get some feedback on whether this approach seems feasible. Is adding the score to OSM data the best approach? Can that information be added as a penalty in the routing calculation easily? Would it need to change something else? Is there an alternative method or approach we should consider?

Additionally, is anyone aware of a similar project or implementation I could reference? Having some guidance or a starting point would be great. 

Any insights, guidance, or points of reference would be greatly appreciated. Thank you for taking the time to read.

Best regards,

**Comment by ImreSamu** [comment]
> The primary audience is women looking to navigate their surroundings
> with an additional layer of safety in mind.
> ...

When setting up a route for female preferences  (on OpenStreetMap) , consider:

Safety & Comfort:
* Prefer routes on well-lit streets in the evening.
* Picking roads with lower car speed limits and fewer lanes.
* Indicate areas with security patrols or frequent police presence for added safety.
* Avoiding areas with dense vegetation or large parks.

Accessibility:
* Prioritize wheelchair-friendly routes, especially when with a stroller or young child.
* Choosing routes with flat sidewalks and fewer steps.
* Prioritize routes with wide sidewalks for easy maneuvering of strollers or wheelchairs.
* Opt for pedestrian zones or areas with minimal vehicle traffic.
* Prioritize paths that have tactile paving for the visually impaired.
* Highlight routes with audible traffic signals for the visually impaired.

Amenities:

* Including public restrooms, toilets.
* Showing drinking water spots.
* Highlighting many points of interest (more POI is better), but exclude/avoiding places like Stadiums or Pubs.
* Highlight routes close to amenities like playgrounds, baby care rooms, or first aid stations.
* Include paths that are near convenience stores or pharmacies for quick access to essentials.

Rest & Weather Protection:

* Prefer paths with benches or rest areas at regular intervals.
* Show routes with shelter or covered paths for protection against weather.

Public Transport:

* Choose routes where buses or other public transport are available.



> Additionally, is anyone aware of a similar project or implementation I could reference?
>  Having some guidance or a starting point would be great.

1.) This is a good research topic, so check Google Scholar : 

* https://scholar.google.com/scholar?q=osm+safety+routing
* https://scholar.google.com/scholar?q=osm+safety+routing+women
* https://scholar.google.com/scholar?q=osm+routing+women
* https://scholar.google.com/scholar?q=Street+Safety+women+routing
* https://scholar.google.com/scholar?q=Street+Female+Safety+route
* .... 

* 
2.)
you can check the GOAT project (  ~ elderly, wheelchair, safe-night  routing ) 
https://github.com/search?q=repo%3Agoat-community%2Fgoat+safe+night&type=code

![image](https://github.com/valhalla/valhalla/assets/2217320/3af8a060-09a1-4b6e-ba4e-2061e303cd7f)

[PDF : Master’s Thesis : Incorporating the four accessibility components into an interactive accessibility instrument](https://mediatum.ub.tum.de/doc/1542813/document.pdf)


3.) 
Data-Driven Approach to Assess Street Safety: Large-Scale Analysis of the Microscopic Design
https://www.mdpi.com/2220-9964/11/11/537 ( see the "References")

**Reply by alz86** [reply]
Hi @ImreSamu thanks for your answer, it gives us many useful insights. Nevertheless, we are looking for specific technical guidance about how to implement our idea with Valhalla, like validating if our approach is the right one, which parts of the system should be modified, etc.

**Reply by ImreSamu** [reply]
@alz86 

I'm not yet clear on the user requirements for the best data model.

For example:

Do we need just one "safety_score" for all transportation modes? 
Or should each mode have its own distinct safety score? For instance, how safe someone feels on the same street can differ if they're in a car or walking.  
![image](https://github.com/valhalla/valhalla/assets/2217320/e530c21c-1f00-4f72-9a2c-73b36cb5a31d)

And some traffic accident statistics consider the mode of transport:
- Casualty: [Cyclist, Pedestrian, Car occupant]
- Severity: [Fatal, Severe, Minor]

for example :
* https://www.cyclestreets.org/sites/bikedata/
   *  [bikedata.cyclestreets.net/collisions](https://bikedata.cyclestreets.net/collisions:field%3Acasualties=Cyclist%2CPedestrian%2CCar%20occupant,photomap/#17/51.511514/-0.103837 )

![image](https://github.com/valhalla/valhalla/assets/2217320/ccae7f36-9d70-4f17-b276-fa3ccbcf6455)

So, I would calculate a different "safety_score" for each mode of transport. This would then be combined with other data found in the Valhalla route graph, like lighting, speed limits, ... 

>   to each street (reported crime rates, presence of street lighting, etc.).

By the way, the Valhalla "graph" already includes street lighting information.
- https://github.com/valhalla/valhalla/pull/3893  ( + https://github.com/valhalla/valhalla/issues/3848 )
  -  https://wiki.openstreetmap.org/wiki/Key:lit

And the "safe" costing option ( for pedestrians) is discussed here:  
* https://github.com/valhalla/valhalla/discussions/3838

**Reply by alz86** [reply]
Hi Imre, thanks again for your help and sorry for my late reply. I will take a look to the links you posted, they are very useful. 

Thanks a lot!

**Comment by kevinkreiser** [comment, marked as answer]
we've been asked this question in one form or another at least 10 times. i think it's about time to write up a proposal for supporting this feature.

at a high level the idea is to allow a person running the routing service to include an 'overlay' layer on top of the graph which can hold a proprietary metric. the routing algorithms costing model and API then need to allow the request to specify whether or not the overlay should be used in edge/node costing and how.

I'll volunteer to write up an official issue about this topic and some options on how we could add support for it.

in the short term we don't have very good options. you could co opt the traffic data or the regular osm speed limit data but you still need to associate that data to osm/graph. you'd also need to modify the highway tags so that the hierarchies play nice with your data or disable hierarchies at request or data building time. it would work but I'd say the effort would be better spent adding first class support for the general feature as a whole

**Reply by alz86** [reply]
Hi Kevin, thanks for your answer, and sorry for my late reply.

It would be great to have a feature like that, I think it will open new possibilities to Valhalla. Will keep an eye on it.

Thanks for your support.

**Comment by ImreSamu** [comment]
related:  GIScience started a new "Safer Route" project:  https://github.com/GIScience/Safer-Route

**Reply by ImreSamu** [reply]
https://heigit.org/heigit-team-wins-open-source-software-for-sustainable-development-goals-oss4sdg-hackathon/
http://lemberg.geog.uni-heidelberg.de:3838/Safer-Route/
